From a4d0e485ab860d8cd950bc81c6b5674e67dd84a9 Mon Sep 17 00:00:00 2001
Message-Id: <a4d0e485ab860d8cd950bc81c6b5674e67dd84a9.1421808898.git.chang-joon.lee@intel.com>
In-Reply-To: <dcec99b5d6fee3bcdab208c48cae68cdc752e9d4.1421808898.git.chang-joon.lee@intel.com>
References: <dcec99b5d6fee3bcdab208c48cae68cdc752e9d4.1421808898.git.chang-joon.lee@intel.com>
From: sudheer gurram <sudheer.gurram@intel.com>
Date: Wed, 3 Dec 2014 13:30:30 +0530
Subject: [PATCH 2/5] MUST_REBASE [VPG]: drm/i915: Fix for the gamma curve
 issue

The following change corrects the IE_MOD_TABLE bit for VLV and CHV
platforms.

MUST_REBASE: DPST can not be upstreamed due to the lack of an
open-source, user-space client.

Issue: GMINL-3320
Change-Id: I30a18d01f5735546c80ac5660b45c0419771ed98
Signed-off-by: sudheer gurram <sudheer.gurram@intel.com>
---
 drivers/gpu/drm/i915/i915_debugfs.c |    2 +-
 drivers/gpu/drm/i915/i915_dpst.c    |   15 +++++++++++----
 drivers/gpu/drm/i915/i915_drv.h     |    1 +
 drivers/gpu/drm/i915/i915_reg.h     |    1 +
 drivers/gpu/drm/i915/i915_suspend.c |    3 ++-
 5 files changed, 16 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_debugfs.c b/drivers/gpu/drm/i915/i915_debugfs.c
index 2cc4313..0e3075e 100644
--- a/drivers/gpu/drm/i915/i915_debugfs.c
+++ b/drivers/gpu/drm/i915/i915_debugfs.c
@@ -1719,7 +1719,7 @@ static int i915_dpst_status(struct seq_file *m, void *unused)
 		   dev_priv->dpst.blc_adjustment * 100 / DPST_MAX_FACTOR);
 
 	seq_printf(m, "IE modification table: %s\n",
-		   blm_hist_ctl & IE_MOD_TABLE_ENABLE ?
+		   blm_hist_ctl & dev_priv->dpst.ie_mod_table_enable ?
 		   "enabled" : "disabled");
 
 	blm_hist_ctl |= BIN_REG_FUNCTION_SELECT_IE;
diff --git a/drivers/gpu/drm/i915/i915_dpst.c b/drivers/gpu/drm/i915/i915_dpst.c
index d0679fa..326287f 100644
--- a/drivers/gpu/drm/i915/i915_dpst.c
+++ b/drivers/gpu/drm/i915/i915_dpst.c
@@ -142,7 +142,8 @@ i915_dpst_disable_hist_interrupt(struct drm_device *dev)
 	/* Disable histogram logic */
 	blm_hist_ctl = I915_READ(dev_priv->dpst.reg.blm_hist_ctl);
 	blm_hist_ctl &= ~IE_HISTOGRAM_ENABLE;
-	blm_hist_ctl &= ~IE_MOD_TABLE_ENABLE;
+	blm_hist_ctl &= ~(dev_priv->dpst.ie_mod_table_enable);
+
 	I915_WRITE(dev_priv->dpst.reg.blm_hist_ctl, blm_hist_ctl);
 
 	/* DPST interrupt in DE_IER register is disabled in irq_uninstall */
@@ -234,7 +235,8 @@ i915_dpst_apply_luma(struct drm_device *dev,
 
 	/* Enable Image Enhancement Table */
 	blm_hist_ctl = I915_READ(dev_priv->dpst.reg.blm_hist_ctl);
-	blm_hist_ctl |= IE_MOD_TABLE_ENABLE | ENHANCEMENT_MODE_MULT;
+	blm_hist_ctl |= dev_priv->dpst.ie_mod_table_enable |
+		ENHANCEMENT_MODE_MULT;
 	I915_WRITE(dev_priv->dpst.reg.blm_hist_ctl, blm_hist_ctl);
 
 	return 0;
@@ -247,7 +249,8 @@ i915_dpst_save_luma(struct drm_device *dev)
 
 	/* Only save if user mode has indeed applied valid settings which
 	 * we determine by checking that the IE mod table was enabled */
-	if (!(I915_READ(dev_priv->dpst.reg.blm_hist_ctl) & IE_MOD_TABLE_ENABLE))
+	if (!(I915_READ(dev_priv->dpst.reg.blm_hist_ctl) &
+				dev_priv->dpst.ie_mod_table_enable))
 		return;
 
 	/* IE mod table entries are saved in the hardware even if the table
@@ -277,7 +280,8 @@ i915_dpst_restore_luma(struct drm_device *dev)
 	/* IE mod table entries are saved in the hardware even if the table
 	 * is disabled, so we only need to re-enable the table */
 	blm_hist_ctl = I915_READ(dev_priv->dpst.reg.blm_hist_ctl);
-	blm_hist_ctl |= IE_MOD_TABLE_ENABLE | ENHANCEMENT_MODE_MULT;
+	blm_hist_ctl |= dev_priv->dpst.ie_mod_table_enable |
+		ENHANCEMENT_MODE_MULT;
 	I915_WRITE(dev_priv->dpst.reg.blm_hist_ctl, blm_hist_ctl);
 }
 
@@ -372,11 +376,13 @@ i915_dpst_update_registers(struct drm_device *dev)
 		dev_priv->dpst.reg.blm_hist_guard = BLM_HIST_GUARD;
 		dev_priv->dpst.reg.blm_hist_bin = BLM_HIST_BIN;
 		dev_priv->dpst.reg.blm_hist_bin_count_mask = BIN_COUNT_MASK_4M;
+		dev_priv->dpst.ie_mod_table_enable = IE_MOD_TABLE_ENABLE;
 	} else if (IS_VALLEYVIEW(dev)) {
 		dev_priv->dpst.reg.blm_hist_ctl = VLV_BLC_HIST_CTL(PIPE_A);
 		dev_priv->dpst.reg.blm_hist_guard = VLV_BLC_HIST_GUARD(PIPE_A);
 		dev_priv->dpst.reg.blm_hist_bin = VLV_BLC_HIST_BIN(PIPE_A);
 		dev_priv->dpst.reg.blm_hist_bin_count_mask = BIN_COUNT_MASK_4M;
+		dev_priv->dpst.ie_mod_table_enable = VLV_IE_MOD_TABLE_ENABLE;
 	} else if (IS_BROADWELL(dev)) {
 		dev_priv->dpst.reg.blm_hist_ctl =
 				BDW_DPST_CTL_PIPE(dev_priv->dpst.pipe);
@@ -386,6 +392,7 @@ i915_dpst_update_registers(struct drm_device *dev)
 				BDW_DPST_BIN_PIPE(dev_priv->dpst.pipe);
 		dev_priv->dpst.reg.blm_hist_bin_count_mask =
 				BIN_COUNT_MASK_16M;
+		dev_priv->dpst.ie_mod_table_enable = IE_MOD_TABLE_ENABLE;
 	} else {
 		DRM_ERROR("DPST not supported on this platform\n");
 		return -EINVAL;
diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 10796a7..f346f48 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -1662,6 +1662,7 @@ struct drm_i915_private {
 		u32 blc_adjustment;
 		u32 gb_delay;
 		u32 init_image_res;
+		u32 ie_mod_table_enable;
 		bool user_enable;	/* user client wishes to enable */
 		bool kernel_disable;	/* kernel override wishes to disable */
 		bool enabled;		/* actual functional state */
diff --git a/drivers/gpu/drm/i915/i915_reg.h b/drivers/gpu/drm/i915/i915_reg.h
index e4c3f12..c2fbcba 100644
--- a/drivers/gpu/drm/i915/i915_reg.h
+++ b/drivers/gpu/drm/i915/i915_reg.h
@@ -6209,6 +6209,7 @@ enum punit_power_well {
 #define BLM_HIST_CTL				0x48260
 #define  IE_HISTOGRAM_ENABLE		(1<<31)
 #define  IE_MOD_TABLE_ENABLE		(1<<27)
+#define  VLV_IE_MOD_TABLE_ENABLE            (1<<30)
 #define  HSV_INTENSITY_MODE			(1<<24)
 #define  ENHANCEMENT_MODE_MULT		(2<<13)
 #define  BIN_REG_FUNCTION_SELECT_IE	(1<<11)
diff --git a/drivers/gpu/drm/i915/i915_suspend.c b/drivers/gpu/drm/i915/i915_suspend.c
index 7b8783b..3ca95b2 100644
--- a/drivers/gpu/drm/i915/i915_suspend.c
+++ b/drivers/gpu/drm/i915/i915_suspend.c
@@ -197,7 +197,8 @@ void i915_save_dpst_regs(struct drm_i915_private *dev_priv)
 
 	/* Disable image enhancement table so we do not apply invalid
 	 * data when we resume */
-	dev_priv->regfile.saveBLM_HIST_CTL &= ~IE_MOD_TABLE_ENABLE;
+	dev_priv->regfile.saveBLM_HIST_CTL &=
+		~(dev_priv->dpst.ie_mod_table_enable);
 }
 
 void i915_restore_dpst_regs(struct drm_i915_private *dev_priv)
-- 
1.7.9.5

